numpy/ma/core.py

```diff
@@ -4287,22 +4298,44 @@ def count(self, axis=None):
         array([3, 0])
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         m = self._mask
-        s = self.shape
-        if m is nomask:
+        # optimization if we exactly have an ndarray
+        if m is nomask and type(self._data) is np.ndarray:
+            # see _count_reduce_items in _methods.py
+            axes = axis
             if axis is None:
-                return self.size
+                axes = tuple(range(self.ndim))
+            if not isinstance(axes, tuple):
+                axes = (axis,)
+
+            items = 1
+            for ax in axes:
+                items *= self.shape[ax]
+
+            if self.shape is () or axis is None:
+                return items
+            elif kwargs.get('keepdims', False):
+                out_dims = list(self.shape)
+                for a in axes:
+                    out_dims[a] = 1
             else:
-                n = s[axis]
-                t = list(s)
-                del t[axis]
-                return np.full(t, n, dtype=np.intp)
-        n1 = np.size(m, axis)
-        n2 = np.sum(m, axis=axis, dtype=np.intp)
-        if axis is None:
-            return (n1 - n2)
-        else:
-            return narray(n1 - n2)
+                out_dims = [d for n,d in enumerate(self.shape) if n not in axes]
+            return np.asarray(np.intp(items)*ones(out_dims, dtype=np.intp))
+
+        # take care of the masked singleton
+        if self is masked:
+            return 0
+
+        # for ndarray subclasses, count should behave exactly the same
```

Comment:
I think this is very risky -- we just want to count masked items, so I think we should stick with counting a regular array by default. In particular, viewing the counts as a `Quantity` will not work (as it will not initialise a unit, which means that down the line, when we divide by the counts, an error will be raised). More generally, I don't think one can assume that for every ndarray subclass boolean type is meaningful!

Now for a matrix, we do have a problem, but I would suggest to explicitly special-case it (I think everybody by now agrees that the different interpretation of shapes by matrix was a bad idea...)

So, specifically, I would suggest removing the lines here and instead adding the following above:
```
m = self._mask
if isinstance(self.data, np.matrix):
    if m is nomask:
        m = np.zeros(self.shape, dtype=np.bool_)
    m = m.view(type(self.data))

if m is nomask:  # i.e., remove isinstance(self.data, np.ndarray) here
...